### PR TITLE
[ENG-738] fix: update encounter end date handling in EncounterQuestion component

### PR DESCRIPTION
**Pull request by abhimanyurajeesh**
## Proposed Changes
This PR fixes an issue with manually updating a encounter after it has been restarted.

Currently, if an encounter end date (discharge date) already exists, it is overwritten with the current date and time. Instead, the existing discharge date should be preserved. If any changes are required, the user can review and update the date manually.

This issue only occurs when the encounter is updated manually through the **update encounter.** It does not affect the **Mark as Complete** flow, as that path handles the encounter correctly. Therefore, this is an edge case specific to manual updates is fixed here.

image

Fixes ENG-738

Tagging: @ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate the bug or test the new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is placed in I18n files.
- [x] Prepare a screenshot or demo video for the changelog entry and attach it to the issue.
- [x] Request peer reviews.
- [ ] Complete QA on mobile devices.
- [ ] Complete QA on desktop devices.
- [ ] Add or update Playwright tests for related changes

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved questionnaire loading and error handling with clearer loading states.
  * Encounter responses now reliably retain existing server or locally entered values.
  * Terminal encounters keep their end date synchronized, with the current date and time used when none is available.
  * Non-terminal encounters clear outdated end dates, while outpatient and discharge details are handled more consistently.
* **Style**
  * Improved text wrapping for hospitalization section headings.

### CODE REVIEW

File: src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx [hunk 1]
```diff
@@ -146,7 +146,7 @@ export function EncounterQuestion({
         handleUpdateEncounter({
           period: {
             ...encounter.period,
-            end: new Date().toISOString(),
+            end: encounterData?.period?.end ?? new Date().toISOString(),
```
[the remark this answers is not in the document]
Comment (abhimanyurajeesh) [reply src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx]: moving it down

File: src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx [hunk 2]
```diff
@@ -385,7 +387,7 @@ export function EncounterQuestion({
       {/* Hospitalization Details - Only show for relevant encounter classes */}
       {["imp", "obsenc", "emer"].includes(encounter.encounter_class) && (
         <div className="col-span-2 border border-gray-200 rounded-lg p-4 space-y-4">
-          <h3 className="text-lg font-semibold break-words">
+          <h3 className="text-lg font-semibold wrap-break-word">
             {t("hospitalization_details")}
           </h3>
```
[the remark this answers is not in the document]
Comment (abhimanyurajeesh) [reply src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx:398]: https://tailwindcss.com/docs/overflow-wrap#wrapping-mid-word

please remember this

File: src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx [hunk 3]
```diff
@@ -239,6 +199,45 @@ export function EncounterQuestion({
...
+        new Date().toISOString())
+      : undefined;
+
+    if (encounter.period.end !== desiredEnd) {
+      handleUpdateEncounter({
+        period: { ...encounter.period, end: desiredEnd },
+      });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [encounter.status]);
```
[the remark this answers is not in the document]
Comment (abhimanyurajeesh) [reply src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx]: edge case not gonna happen

**Review by nihal467 — APPROVED** [review]

File: src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx [hunk 4]
```diff
@@ -239,6 +199,45 @@ export function EncounterQuestion({
     );
   };
 
+  useEffect(() => {
```
Comment (Jacobjeevan) [inline src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx]: handleUpdateEncounter is going to be called anyways on status change, so we can really move this logic into handleUpdateEncounter itself and remove the useEffect.

**Review by Jacobjeevan — COMMENTED** [review]
LGTM, minor comment.

One thing worth considering is bypassing the mark for discharge button when discharge date is already present.

**Review by Jacobjeevan — APPROVED** [review]

File: src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx [hunk 5]
```diff
@@ -66,12 +68,85 @@ interface EncounterQuestionProps {
...
+  server: EncounterRead | undefined,
+): Hospitalization {
+  if (OUTPATIENT_ENCOUNTER_CLASSES.includes(encounter.encounter_class)) {
+    return {};
+  }
+  return {
+    ...encounter.hospitalization,
+    discharge_disposition: resolveDischargeDisposition(encounter, server),
+  };
+}
```
Comment (rithviknishad) [inline src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx:120]: `validateEncounterQuestion` is using in-patient or similar array for validating hospitalization discharge disposition.

you could specify `hospitalization_encounter_classes` and re-use it in both places.

File: src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx [hunk 6]
```diff
@@ -66,12 +68,85 @@ interface EncounterQuestionProps {
...
+
+function resolvePeriodEnd(
+  encounter: EncounterEdit,
+  server: EncounterRead | undefined,
```
Comment (rithviknishad) [inline src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx]:
```suggestion
  edit: EncounterEdit,
  encounter: EncounterRead | undefined,
```

File: src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx [hunk 7]
```diff
@@ -66,12 +68,85 @@ interface EncounterQuestionProps {
...
+  // Leaving a terminal status clears the end date, so fall back to the
+  // recorded one rather than stamping "now" on the way back in.
+  return (
+    encounter.period.end ?? server?.period?.end ?? new Date().toISOString()
+  );
+}
+
+function normalizeEncounter(
+  encounter: EncounterEdit,
+  server: EncounterRead | undefined,
```
Comment (rithviknishad) [inline src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx]: here as wwll

File: src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx [hunk 8]
```diff
@@ -122,133 +197,56 @@ export function EncounterQuestion({
...
-  });
+  // The questionnaire response is the single source of truth for the encounter
+  // being edited; a copy in local state only ever drifts out of sync with it.
+  const encounter = (
+    questionnaireResponse.values[0]?.value as EncounterEdit[] | undefined
+  )?.[0];
 
+  // Seed the response from the server once per encounter. Skipped if the form
+  // already has an answer (e.g. a restored draft) so it isn't clobbered on remount.
+  const seededEncounterId = useRef<string>(null);
```
Comment (rithviknishad) [inline src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx]: i don't understand what `seedEncounterId` is used for...

**Review by rithviknishad — CHANGES_REQUESTED** [review]

File: src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx [hunk 9]
```diff
@@ -66,12 +68,87 @@ interface EncounterQuestionProps {
...
+function resolvePeriodEnd(
+  edit: EncounterEdit,
+  encounter: EncounterRead | undefined,
+): string | undefined {
+  if (!TERMINAL_ENCOUNTER_STATUSES.includes(edit.status)) {
+    return undefined;
+  }
+  // Leaving a terminal status clears the end date, so fall back to the
+  // recorded one rather than stamping "now" on the way back in.
+  return edit.period.end ?? encounter?.period?.end ?? new Date().toISOString();
```
[the remark this answers is not in the document]
Comment (abhimanyurajeesh) [reply src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx]: expected

File: src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx [hunk 10]
```diff
@@ -395,9 +389,11 @@ export function EncounterQuestion({
       )}
 
       {/* Hospitalization Details - Only show for relevant encounter classes */}
-      {["imp", "obsenc", "emer"].includes(encounter.encounter_class) && (
+      {HOSPITALIZATION_ENCOUNTER_CLASSES.includes(
+        encounter.encounter_class,
+      ) && (
         <div className="col-span-2 border border-gray-200 rounded-lg p-4 space-y-4">
-          <h3 className="text-lg font-semibold break-words">
+          <h3 className="text-lg font-semibold wrap-break-word">
```
[the remark this answers is not in the document]
Comment (abhimanyurajeesh) [reply src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx:396]: i just fixed what was broken

File: src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx [hunk 11]
```diff
@@ -158,7 +158,7 @@ export function EncounterQuestion({
         handleUpdateEncounter({
```
[the remark this answers is not in the document]
Comment (abhimanyurajeesh) [reply src/components/Questionnaire/QuestionTypes/EncounterQuestion.tsx]: Not from this PR
